### PR TITLE
Flag dataset item coords as readonly

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Changing the behavior of the dataset item coord dicts to be readonly:
```python
del ds['a'].coords['x']
```

now raises instead of doing nothing. This is then consistent with slices:

```python
del da['x', 17].coords['x']   # raises already
```

You can shallow-copy to clear the readonly flags.
```python
a = ds['a'].copy(deep=False)
del a.coords['x']   # works
```

Note that this change exposes some "bugs" that krept into our unit tests as part of the recent `DataArray` and `Dataset` refactor. A number of tests was adding coords to items of datasets, which previously set item attrs, whereas after the rewrite it silently dropped them.

This was discovered as part of a small dataset-test cleanup, so a number of unrelated cleanup/refactor is contained (see individual earlier commits).

### CODE REVIEW

File: dataset/shape.cpp [hunk 1]
```diff
@@ -101,11 +101,11 @@ Dataset concatenate(const Dataset &a, const Dataset &b, const Dim dim) {
...
-  auto result =
-      a.empty()
-          ? Dataset(std::map<std::string, Variable>(),
-                    concat(a.coords(), b.coords(), dim, a.sizes(), b.sizes()))
-          : Dataset();
+  Dataset result;
+  if (a.empty())
+    result.setCoords(
+        Coords(concatenate(a.sizes(), b.sizes(), dim),
+               concat(a.coords(), b.coords(), dim, a.sizes(), b.sizes())));
```
Comment (jl-wynen) [inline dataset/shape.cpp:108]: Not the point of this PR, but why do we have `concatenate` and `concat`?

File: dataset/test/dataset_arithmetic_test.cpp [hunk 2]
```diff
@@ -719,39 +719,6 @@ TYPED_TEST(DatasetBinaryOpTest, masks_not_shared) {
...
-  auto d = copy(dense.slice({Dim::Y, 0}));
-  dense.setData("data_x_1", dense["data_x"]);
-  EXPECT_EQ(dense["data_x"], dense["data_x_1"]);
-
-  d.setCoord(Dim("l1"),
-             makeVariable<double>(Dims{Dim::X},
-                                  Shape{d.coords()[Dim::X].dims().volume()}));
-  dense.setData("data_x_2", d["data_x"]);
-  EXPECT_TRUE(dense.coords().contains(Dim("l1")));
-}
```
Comment (jl-wynen) [inline dataset/test/dataset_arithmetic_test.cpp:753]: Why did you delete those tests?

File: dataset/test/dataset_test.cpp [hunk 3]
```diff
@@ -158,6 +157,48 @@ TEST(DatasetTest, setData_with_and_without_variances) {
...
+  // Metadata *dicts* are not shared
+  ds.coords().erase(Dim::X);
+  EXPECT_NE(ds[name].coords(), array.coords());
+  EXPECT_TRUE(array.coords().contains(Dim::X));
+  ds[name].masks().erase("mask");
+  EXPECT_NE(ds[name].masks(), array.masks());
+  EXPECT_TRUE(array.masks().contains("mask"));
+  ds[name].attrs().erase(Dim("attr"));
+  EXPECT_NE(ds[name].attrs(), array.attrs());
+  EXPECT_TRUE(array.attrs().contains(Dim("attr")));
```
Comment (jl-wynen) [inline dataset/test/dataset_test.cpp:180]: Is there a public interface for comparing the VariableConcepts?

File: dataset/shape.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply dataset/shape.cpp:108]: `concat` is an implementation detail and not user-exposed.

File: dataset/test/dataset_arithmetic_test.cpp (hunk 2 above)
Comment (SimonHeybrock) [reply dataset/test/dataset_arithmetic_test.cpp:753]: I first moved them to `dataset_test.cpp` where they belong, and then subsequently removed/replaced them by more modern and complete tests.

File: dataset/test/dataset_test.cpp (hunk 3 above)
Comment (SimonHeybrock) [reply dataset/test/dataset_test.cpp:180]: No there is not. Which aspect are you referring to in this context?

Comment (jl-wynen) [reply dataset/test/dataset_test.cpp:180]: It is a bit awkward to test if two objects share memory by modifying one and checking if the other changed. It would be nicer if we could ask them directly if the buffers are the same.

Comment (SimonHeybrock) [reply dataset/test/dataset_test.cpp:180]: There is `Variable::is_same` which was added recently, it is just not used widely yet.

**Review by jl-wynen — APPROVED** [review]
